kratos/includes/geometry_container.h

```diff
@@ -0,0 +1,407 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Pooyan Dadvand
+//
+//
+
+
+#if !defined(KRATOS_GEOMETRY_CONTAINER_H_INCLUDED )
+#define  KRATOS_GEOMETRY_CONTAINER_H_INCLUDED
+
+
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "containers/pointer_hash_map_set.h"
+#include "containers/data_value_container.h"
+
+
+namespace Kratos
+{
+
+///@name Kratos Classes
+///@{
+
+template<class TGeometryType>
+class GeometryContainer : public DataValueContainer, public Flags
+{
+    class GetGeometryId : public std::unary_function<const TGeometryType* const, std::size_t>
+    {
+    public:
+        std::size_t const& operator()(const TGeometryType& rGeometry) const
+        {
+            return rGeometry.Id();
+        }
+    };
+
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of GeometryContainer
+    KRATOS_CLASS_POINTER_DEFINITION(GeometryContainer);
+
+    typedef std::size_t IndexType;
+    typedef std::size_t SizeType;
+
+    /* Geometry Hash Map Container.
+    *  Hash of Id are keys to corresponding intrusive pointer */
+    typedef PointerHashMapSet<
+        TGeometryType,
+        std::hash<std::size_t>,
+        GetGeometryId,
+        typename TGeometryType::Pointer
+    > GeometriesContainerType;
+
+    /// Geometry Iterator
+    typedef typename GeometriesContainerType::iterator GeometryIterator;
+
+    /// Const Geometry Iterator
+    typedef typename GeometriesContainerType::const_iterator GeometryConstantIterator;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default Constructor
+    GeometryContainer()
+        : Flags()
+        , mpGeometries(new GeometriesContainerType())
```

Comment:
This looks like a mem-leak, I don't see it destructed somewhere

Reply:
It's a smart pointer. I think it does not need to be destructed.